cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -651,9 +667,33 @@ func (s *service) send(evt interface{}) {
 	s.events <- evt
 }
 
+func (s *service) handleProcessExitL(e runcC.Exit, c *runc.Container, p process.Process) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.handleProcessExit(e, c, p)
+}
+
 // s.mu must be locked when calling handleProcessExit
 func (s *service) handleProcessExit(e runcC.Exit, c *runc.Container, p process.Process) {
 	if ip, ok := p.(*process.Init); ok {
+		s.pendingExecsLock.Lock()
+		wg, pendingExecs := s.pendingExecs[p.Pid()]
+		s.pendingExecsLock.Unlock()
+		if pendingExecs {
+			// This is the init process and there are pending execs to process, so
+			// wait until all pending execs are processed before we process our exit.
+			// see: https://github.com/containerd/containerd/issues/9719
+			go func() {
+				wg.Wait()
```

Comment:
I'm not entirely sold on the use of goroutines and WaitGroups for this. The `handleStarted` closure which wakes this goroutine with `wg.Done()` could instead check that it is the last pending exec for an exited init process and synchronously dispatch the init process' TaskExit event. (Is the use of goroutines and WaitGroups the reason why `pendingExecs` can't be guarded with `lifecycleMu`?) 

The `handleStarted` closure knows the init process' PID and is subscribed to exit events already; it could trivially check if the init process had exited while the exec was starting. The only new information it needs to know is whether or not it is handling an early exit for _the last remaining_ exec in the container. Similarly, `processExit` could use the same information to know to suppress the `handleProcessExit` call for an init process with outstanding execs. That could be implemented e.g. by adding a counter field to the struct currently named `containerProcess`. I think it's feasible to fix exit ordering without any new goroutines, mutexes, or `service` struct fields, by leveraging the existing `exitSubscribers` and `running` fields.

Reply:
> The `handleStarted` closure knows the init process' PID and is subscribed to exit events already; it could trivially check if the init process had exited while the exec was starting

But the problem isn't letting the `handleStarted` closure check if the init process has exited while the exec is starting, the problem is knowing whether to delay emitting the init process exit or not/know whether there might be pending execs.

> Similarly, `processExit` could use the same information to know to suppress the handleProcessExit call for an init process with outstanding execs. That could be implemented e.g. by adding a counter field to the struct currently named containerProcess.

That might be viable, I guess we'd increment `containerProcessPendingExecs` before starting and decrement after processing the exec, and we could use that in `processExits` to know whether we can process that exit or not. However, I don't understand how we can do that `without any new goroutines, ...`. We have to launch a goroutine for that init process exit so we can block until it can be processed and continue processing other exits in the meantime, no? Otherwise we have to make severe modifications to this code. Another question is: if we remove the `WaitGroup`, will we hand craft some implementation using channels/a `sync.Cond`/whatever to let our waiting init process exits that they should exit, or do something else entirely?